virtual_rainforest/core/axes.py

```diff
@@ -140,12 +140,21 @@ def get_validator(axis: str, data: Data, darray: DataArray) -> Optional[Callable
 
     This function iterates over the registered validator functions for a given core axis
     in the AXIS_VALIDATORS registry and matches the data array signatures against the
-    provided signatures. It returns either the appropriate function or None.
+    provided signatures.
+
+    If the input data array uses dimension names that have been registered to validate a
+    particular axis, then a matching validator must be found and an exception is raised
+    if no signature is matched. If the dimension names on the data array do not match
+    any of the registered dimensions, then the function returns None.
 
     Args:
         axis: The core axis to get a validator for
         data: A Data object providing parameter information
```

Comment:
Given that you're only using `grid_type` I would pass only that as argument and not the whole `data` object. You say in the doctoring above that in the future, some validators might use other information contained in the `data` object, but given that `data` is not passed to the validators themselves, only used in `get_validator`, there's no point on passing the whole `data` to this function. That will also save you from importing `Data` here, removing the potential issue of circular imports.

Reply:
Actually, `data` is passed on to the validators and `data.grid` is used within the spatial validators. At the moment though, using `Data` is a complicated way to pass in a `Grid` object. 

I think we'll ultimately need a `Config` class. Then essentially all information needed to configure axis validators (and anything else) will be accessed by passing that around? For the moment, just using the `Grid` rather than the `Data` seems more sensible.

Reply:
I don't have the big picture in my head, but I'm not sure that creating another class is the way to go. You already have the configuration dictionary named `config`, so having a `Config` class will make things confusing. 

About `Data`, it only has `grid` and `data` attributes, so I'm not sure what else a validator might use. If you're referring to the `attrs` of the `data` Dataset, I'd pass those as `kwargs` and let each validator pick whatever they need. The same applies to other dimensions or coordinates already contained in the `data` Dataset. Ideally, you want to include as arguments to functions the information actually used in the function, and not big, complex objects that contain the information used. 

So, for the validators you have now, I'd rather use:

```python
def some_spacial_validator(data: DataArray, grid: Grid, **kwargs) -> DataArray:
    ...
```

And if you ever get a validator that require something else, like `param_1` contained in `data.attrs`, in addition to `grid`:

```python
def another_validator(data: DataArray, grid: Grid, param_1: float, **kwargs) -> DataArray:
    ...
```

And if you have validators that have nothing to do with the grid, but just some existing coordinate in the `data` dataset, for example `time`, then:

```python
def a_time_validator(data: DataArray, time: DataArray, **kwargs) -> DataArray:
    ...
```

All of them being called as:

```python
axis_validator_func(value, grid = self.grid, **dict(self.data.data_vars), **dict(self.data.coords), **self.data.attrs)
```

If we follow the variable names used in `data.py:170`.

As cumbersome as this call might look, it is cheap because all in there is passed as reference, it needs to be written only once and, more importantly, it let each validator to be explicit on what it actually needs to run, much better and clear to document.

Reply:
I was thinking longer term about _replacing_ the `config` dict with a class. At the moment the config process uses a chain of functions which  pass information back and forth between variables - most of that is about updating the state of the config dict and the schema that should be applied to it. The code would be simpler and cleaner if we had a `UserDict` style class which had methods updating attributes. That class would also retain more information about the config process - files used etc.

But - that's for a different branch 😄